File: APP/ios/AirHealthApp/App/HistoryViewState.swift
import Foundation

private let oralBaselineRequiredSessions = 3

struct OralHistoryViewItem {
    let sessionID: String
    let title: String
    let detail: String
    let syncLabel: String
    let progressLabel: String
}

struct OralHistorySurfaceState {
    let baselineProgressLabel: String
    let progressDetail: String
    let latestStatusLabel: String?
    let items: [OralHistoryViewItem]
}

struct FatHistoryViewItem {
    let sessionID: String
    let title: String
    let detail: String
    let syncLabel: String
    let finalDeltaLabel: String
    let bestDeltaLabel: String
}

struct FatHistorySurfaceState {
    let latestFinalDeltaLabel: String?
    let bestDeltaLabel: String?
    let pendingCount: Int
    let syncedCount: Int
    let items: [FatHistoryViewItem]
}

extension SessionHistoryStoreState {
    func oralHistorySurface() -> OralHistorySurfaceState {
        let oralRecords = records
            .filter { $0.feature == .oralHealth }
            .sorted { $0.recordedAtEpochMillis > $1.recordedAtEpochMillis }

        let chronologicalProgress = Dictionary(
            uniqueKeysWithValues: oralRecords
                .sorted { $0.recordedAtEpochMillis < $1.recordedAtEpochMillis }
                .enumerated()
                .map { index, record in
                    let label = index < oralBaselineRequiredSessions
                        ? "Baseline session \(index + 1) of \(oralBaselineRequiredSessions)"
                        : "Post-baseline trend check"
                    return (record.sessionID, label)
                }
        )

        let items = oralRecords.map { record in
            OralHistoryViewItem(
                sessionID: record.sessionID,
                title: record.summaryTitle,
                detail: record.summaryDetail,
                syncLabel: record.syncState.displayLabel,
                progressLabel: chronologicalProgress[record.sessionID] ?? "Baseline session"
            )
        }

        let completedCount = min(oralRecords.count, oralBaselineRequiredSessions)
        let baselineProgressLabel = completedCount < oralBaselineRequiredSessions
            ? "Baseline building \(completedCount)/\(oralBaselineRequiredSessions)"
            : "Baseline ready"
        let progressDetail = completedCount < oralBaselineRequiredSessions
            ? "\(oralBaselineRequiredSessions - completedCount) more oral sessions are needed before post-baseline trend views unlock."
            : "Baseline is established. New oral sessions now extend the post-baseline trend view."

        return OralHistorySurfaceState(
            baselineProgressLabel: baselineProgressLabel,
            progressDetail: progressDetail,
            latestStatusLabel: items.first?.syncLabel,
            items: items
        )
    }

    func fatHistorySurface() -> FatHistorySurfaceState {
        let fatRecords = records
            .filter { $0.feature == .fatBurning }
            .sorted { $0.recordedAtEpochMillis > $1.recordedAtEpochMillis }

        var runningBestDelta = 0
        let items = fatRecords
            .sorted { $0.recordedAtEpochMillis < $1.recordedAtEpochMillis }
            .map { record -> FatHistoryViewItem in
                let finalDelta = deriveFinalDelta(from: record.resultToken)
                runningBestDelta = max(runningBestDelta, finalDelta)
                return FatHistoryViewItem(
                    sessionID: record.sessionID,
                    title: record.summaryTitle,
                    detail: record.summaryDetail,
                    syncLabel: record.syncState.displayLabel,
                    finalDeltaLabel: "Final delta +\(finalDelta)",
                    bestDeltaLabel: "Best delta +\(runningBestDelta)"
                )
            }
            .reversed()

        return FatHistorySurfaceState(
            latestFinalDeltaLabel: items.first?.finalDeltaLabel,
            bestDeltaLabel: items.map(\.bestDeltaLabel).max(by: { labelScore($0) < labelScore($1) }),
            pendingCount: fatRecords.filter { $0.syncState == .pending }.count,
            syncedCount: fatRecords.filter { $0.syncState == .synced }.count,
            items: Array(items)
        )
    }
}

private extension SessionSyncState {
    var displayLabel: String {
        switch self {
        case .pending:
            return "Pending sync"
        case .synced:
            return "Synced"
        }
    }
}

private func deriveFinalDelta(from resultToken: String) -> Int {
    resultToken.unicodeScalars.reduce(0) { accumulator, scalar in
        ((accumulator * 31) + Int(scalar.value)) % 21
    } + 5
}

private func labelScore(_ label: String) -> Int {
    Int(label.split(separator: "+").last ?? "0") ?? 0
}

File: APP/android/app/src/main/java/com/airhealth/app/HistoryViewState.kt
package com.airhealth.app

private const val ORAL_BASELINE_REQUIRED_SESSIONS = 3

data class OralHistoryViewItem(
    val sessionId: String,
    val title: String,
    val detail: String,
    val syncLabel: String,
    val progressLabel: String,
)

data class OralHistorySurfaceState(
    val baselineProgressLabel: String,
    val progressDetail: String,
    val latestStatusLabel: String?,
    val items: List<OralHistoryViewItem>,
)

data class FatHistoryViewItem(
    val sessionId: String,
    val title: String,
    val detail: String,
    val syncLabel: String,
    val finalDeltaLabel: String,
    val bestDeltaLabel: String,
)

data class FatHistorySurfaceState(
    val latestFinalDeltaLabel: String?,
    val bestDeltaLabel: String?,
    val pendingCount: Int,
    val syncedCount: Int,
    val items: List<FatHistoryViewItem>,
)

fun SessionHistoryStoreState.oralHistorySurface(): OralHistorySurfaceState {
    val oralRecords = records
        .filter { it.feature == FeatureKind.ORAL_HEALTH }
        .sortedByDescending { it.recordedAtEpochMillis }

    val chronologicalProgressLabels = oralRecords
        .sortedBy { it.recordedAtEpochMillis }
        .mapIndexed { index, record ->
            record.sessionId to if (index < ORAL_BASELINE_REQUIRED_SESSIONS) {
                "Baseline session ${index + 1} of $ORAL_BASELINE_REQUIRED_SESSIONS"
            } else {
                "Post-baseline trend check"
            }
        }
        .toMap()

    val items = oralRecords.map { record ->
        OralHistoryViewItem(
            sessionId = record.sessionId,
            title = record.summaryTitle,
            detail = record.consumerSafeSummaryDetail(),
            syncLabel = record.syncState.displayLabel(),
            progressLabel = chronologicalProgressLabels.getValue(record.sessionId),
        )
    }

    val completedCount = oralRecords.size.coerceAtMost(ORAL_BASELINE_REQUIRED_SESSIONS)
    val baselineProgressLabel = if (completedCount < ORAL_BASELINE_REQUIRED_SESSIONS) {
        "Baseline building $completedCount/$ORAL_BASELINE_REQUIRED_SESSIONS"
    } else {
        "Baseline ready"
    }
    val progressDetail = if (completedCount < ORAL_BASELINE_REQUIRED_SESSIONS) {
        "${ORAL_BASELINE_REQUIRED_SESSIONS - completedCount} more oral sessions are needed before post-baseline trend views unlock."
    } else {
        "Baseline is established. New oral sessions now extend the post-baseline trend view."
    }

    return OralHistorySurfaceState(
        baselineProgressLabel = baselineProgressLabel,
        progressDetail = progressDetail,
        latestStatusLabel = items.firstOrNull()?.syncLabel,
        items = items,
    )
}

fun SessionHistoryStoreState.fatHistorySurface(): FatHistorySurfaceState {
    val fatRecords = records
        .filter { it.feature == FeatureKind.FAT_BURNING }
        .sortedByDescending { it.recordedAtEpochMillis }

    var runningBestDelta = 0
    val items = fatRecords
        .sortedBy { it.recordedAtEpochMillis }
        .map { record ->
            val finalDelta = deriveFinalDelta(record.resultToken)
            runningBestDelta = maxOf(runningBestDelta, finalDelta)
            record to FatHistoryViewItem(
                sessionId = record.sessionId,
                title = record.summaryTitle,
                detail = record.consumerSafeSummaryDetail(),
                syncLabel = record.syncState.displayLabel(),
                finalDeltaLabel = "Final delta +$finalDelta",
                bestDeltaLabel = "Best delta +$runningBestDelta",
            )
        }
        .reversed()
        .map { it.second }

    return FatHistorySurfaceState(
        latestFinalDeltaLabel = items.firstOrNull()?.finalDeltaLabel,
        bestDeltaLabel = items.maxByOrNull { deriveLabelScore(it.bestDeltaLabel) }?.bestDeltaLabel,
        pendingCount = fatRecords.count { it.syncState == SessionSyncState.PENDING },
        syncedCount = fatRecords.count { it.syncState == SessionSyncState.SYNCED },
        items = items,
    )
}

private fun SessionSyncState.displayLabel(): String {
    return when (this) {
        SessionSyncState.PENDING -> "Pending sync"
        SessionSyncState.SYNCED -> "Synced"
    }
}

private fun deriveFinalDelta(resultToken: String): Int {
    return resultToken.fold(0) { accumulator, character ->
        ((accumulator * 31) + character.code) % 21
    } + 5
}

private fun deriveLabelScore(label: String): Int {
    return label.substringAfter("+").toIntOrNull() ?: 0
}
